### PR TITLE
Create nested items correctly

### LINKED ISSUE #1812: Some Radio Buttons seems to deselect automatically after selection.
**Describe the bug**
- Some radio buttons deselect automatically when selected. 
- Attached are the questions with such behaviors. 
[new_pregnancy_registration.json.zip](https://github.com/google/android-fhir/files/10427105/new_pregnancy_registration.json.zip)
- This happens with the latest SDK code on main

**Component**
SDC library

**To Reproduce**
Load the attached form to the SDK then test using the steps shown on the attached video
https://user-images.githubusercontent.com/4817531/212714001-5b944d84-d6ed-4a6a-9927-f892ac73a9f8.mov

**Expected behavior**
- [ ] The selections on a radio button should only change if the radio button is clicked on and not automatically. 

**Smartphone (please complete the following information):**
 - Device: Xiaomi 10T pro
 - Android version: Android 11

### PR DESCRIPTION
**Pull request by jingtang10**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1812 #1784 #487

**Description**
- Correct the logic to generate questionnaire item view items for nested items
- Improve documentation for a number of existing functions
- Stop creating unnecessary new questionnaire response items during questionnaire state generation

**Alternative(s) considered**
NA

**Type**
Bug fix

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaireItemComponents.kt [hunk 1]
```diff
@@ -424,6 +416,33 @@ internal val Questionnaire.QuestionnaireItemComponent.sliderStepValue: Int?
...
 
+/**
+ * Whether the corresponding [QuestionnaireResponse.QuestionnaireResponseItemComponent] should have
+ * [QuestionnaireResponse.QuestionnaireResponseItemComponent]s nested under
+ * [QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent]s.
+ *
+ * This is true for the following two cases:
+ * 1. Questions with nested items
+ * 2. Repeated groups with nested items (Note that this is how repeated groups are organized in the
+ * QuestionnaireViewModel, and that they will be flattened in the final QuestionnaireResponse.)
```
Comment (omarismail94) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaireItemComponents.kt:427]:
```suggestion
 * [QuestionnaireViewModel], and that they will be flattened in the final [QuestionnaireResponse].)
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 2]
```diff
@@ -630,31 +619,42 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
...
-            // after the question.
-            questionnaireItem.repeats -> questionnaireResponseItem.answer.map { it.item }
-            // Non-repeated questions may have nested items, which we should display
-            else -> listOf(questionnaireResponseItem.item)
-          }
-        nestedResponses.forEach { nestedResponse ->
+      )
+
+      // Add nested questions after the parent item. We need to get the questionnaire items and the
+      // matching questionnaire response items (there could be multiple sets of them) and generate
```
Comment (omarismail94) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:639]: can you re-word as what `them`  is referring to is ambiguous

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/ValueConstraintExtensionValidator.kt [hunk 3]
```diff
@@ -33,7 +33,7 @@ internal open class ValueConstraintExtensionValidator(
     answers: List<QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent>,
     context: Context
   ): ConstraintValidator.ConstraintValidationResult {
-    if (questionnaireItem.hasExtension(url) && !answers.isEmpty()) {
+    if (questionnaireItem.hasExtension(url) && answers.isNotEmpty()) {
       val extension = questionnaireItem.getExtensionByUrl(url)
       // TODO(https://github.com/google/android-fhir/issues/487): Validates all answers.
```
Comment (omarismail94) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/validation/ValueConstraintExtensionValidator.kt:38]: remove this TODO as we're closing this issue with the PR?

**Review by omarismail94 — APPROVED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/ValueConstraintExtensionValidator.kt (hunk 3 above)
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/validation/ValueConstraintExtensionValidator.kt:38]: we're not closing this issue - i will send a follow up pr

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt (hunk 2 above)
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt]: done

**Pull request merged**